### PR TITLE
Change Test Structure for app/api/dao/task_comment.py

### LINKED ISSUE #553: Change Test Structure for app/api/dao/task_comment.py
## Description
I want to change the structure of the /tests/task_comments/test_dao.py tests so as to follow the test structure of the rest of the tests in the project. 
The link to this file is this:
https://github.com/anitab-org/mentorship-backend/blob/develop/tests/task_comments/test_dao.py 

Each method should be tested in a specific class with different methods testing each test case. For example, the file for testing the method delete_comment should be named tests_dao_task_comment_delete_comment.

## Test Cases
The required test cases addressed for this issue are already implemented but should be structured in a different way mentioned in the description.

## Definition of Done
- [ ] All related test cases structured differently should be passing.
- [ ] Approval by 1 mentor.

## Estimation
2 hours

### PR DESCRIPTION
**Pull request by EfthymiaKostaki**
### Description

This pull request changes the test structure for the dao testing of app/api/dao/task_comment.py. More specifically, instead of using the file tests/task_comments/test_dao.py, I created a test base which creates in the setup a task_comment since it was used for testing in all the methods of task_comment.py. I separated the test_dao.py file into four different files to follow the structure of tests followed in the other tests in the project. So each new file tests a specific method of task_comment.py, so when further tests are written to maintain the project they won't be mixed up in one file for all the methods.

Fixes #553

### Type of Change:

- Quality Assurance

**Code/Quality Assurance Only**
- Test refactor

### How Has This Been Tested?

All tests in new files passing. These are the same tests that were tested before.

### Checklist:

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials

**Code/Quality Assurance Only**

- [x] My changes generate no new warnings 
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by vj-codes** [comment]
@EfthymiaKostaki Can u pls tag a maintainer to review this issue?

**Comment by EfthymiaKostaki** [comment]
@vj-codes sure!
@isabelcosta I am looking forward to your review when available!

File: tests/task_comments/task_comments_base_setup.py
Comment (SanketDG) [inline tests/task_comments/task_comments_base_setup.py:3]: Please remove the unused imports from here :smile:

File: tests/task_comments/test_dao_modify_comment.py
Comment (SanketDG) [inline tests/task_comments/test_dao_modify_comment.py:20]: There seems to be a scope for imporvement where you can also check if the `modified_date` is later than `creation_date`

https://github.com/anitab-org/mentorship-backend/blob/80114acd5937768f7680f9ac5542282127021c90/app/database/models/task_comment.py#L28

File: tests/task_comments/test_dao_modify_comment.py
Comment (SanketDG) [inline tests/task_comments/test_dao_modify_comment.py]: Same `assert`s twice? :thinking:

**Review by SanketDG — CHANGES_REQUESTED** [review]
This looks great.

I am not sure if it's needed exactly. I understand you would not have "too" many tests in one file, but 6 is not a a large number. Infact, this makes the developer have to navigate to another file to understand what is being "setup". [The Python Docs also suggest this](https://docs.python.org/3/library/unittest.html#organizing-test-code)

But if these accepted conventions in the project, I am completely fine with it :+1: 

---
New here, I may not be aware of some things that are already communicated.

**Comment by isabelcosta** [comment]
@SanketDG thank you for your input here, your resource was insightful :)
We do have to be careful about over "engineering" / "architecting". What would you suggest then keep these tests into 1 file how it was? I am really interested in another opinion on this.

**Comment by SanketDG** [comment]
@isabelcosta 

> What would you suggest then keep these tests into 1 file how it was?

Yes, because that's how the rest of the project structures its tests.

If we know for sure, that maintaining a lot of tests in a single file will turn into a problem, it makes sense to move to separate files. But they are not a problem, right now.

I would also love to hear what @anitab-org/coding-team thinks about this!

**Comment by EfthymiaKostaki** [comment]
@SanketDG @isabelcosta 
Hey! Thank you both for your input.
@SanketDG I haven't addressed your suggested changes yet because I was waiting for @isabelcosta view on them and the pull request.

I understand your points but I wanted to explain to you my opinion regarding this as well!

1.  I started working on this taking into account the structure of the tests in the other parts of the software. 
I saw that tasks and mentorship relation folder was structured like this:
image
image

2.  Of course, I saw in other parts of the project in the testing this structure wasn't followed, but the main idea was the consistency of the testing material so as everyone that needs to write some testing (for example to test a new method) will immediately understand where he/she should do so.
image

3. Expandability of the testing. 
I think this may overlap with the previous point but for example, if we realize at some point that we need on the task comments further testing on the DAO part or if new functionality is added then, according to the other parts of the testing I highlighted in point 1, we shouldn't use the same class for testing the same method more than once and also test the other methods.

4. Base test for setup
I saw that all the methods in task_comments_dao were using the same method for the setup that is why I created a test base for the task comments so to avoid code repetition. My thinking is that the setup test base should be further expanded so the task_comments_api classes who test each method use this as well since they all have a setup and some common code. With the change suggested in this pull request, me or someone else interested would be able to intergrade these changes I just mentioned so that the test base covers both the API and the DAO part of the task comments.

It is the first time I am working on such principles so it is possible that there are parts of this software I have not understood well and I am eager to hear your input on this and how I should proceed.

File: tests/task_comments/test_dao_modify_comment.py
Comment (EfthymiaKostaki) [reply tests/task_comments/test_dao_modify_comment.py]: I changed all the files without altering anything in the code because in this PR I suggested changing the test structure. If needed I will change these lines in this PR too.

File: tests/task_comments/task_comments_base_setup.py
Comment (EfthymiaKostaki) [reply tests/task_comments/task_comments_base_setup.py:3]: Thanks! I will do that in the next commit.

File: tests/task_comments/test_dao_modify_comment.py
Comment (EfthymiaKostaki) [reply tests/task_comments/test_dao_modify_comment.py:20]: Interesting point! I hadn't thought of that so probably with adding a new test the task_comments_dao would have to be split anyway, right? This could be a new issue.

File: tests/task_comments/task_comments_base_setup.py
Comment (SanketDG) [inline tests/task_comments/task_comments_base_setup.py]: Are we still supporting Python 2? We should drop it even if we are.

If we are just supporting Python 3, this can just be:

```
super()
```

Feel free to open a new issue for this and the others

**Comment by SanketDG** [comment]
> I saw that tasks and mentorship relation folder was structured like this:

>  I saw in other parts of the project in the testing this structure wasn't followed

After taking a second look at `tasks.tests` this seems like a step in the right direction for the sake of extendibility.

I am okay with this change. :rocket:

File: tests/task_comments/task_comments_base_setup.py
Comment (bartekpacia) [reply tests/task_comments/task_comments_base_setup.py]: AFAIK this repository is Python3-only.

Comment (EfthymiaKostaki) [reply tests/task_comments/task_comments_base_setup.py]: Hey, I will open an issue for that.

**Comment by EfthymiaKostaki** [comment]
@SanketDG 
@isabelcosta 
Hey, I fixed the issues you mentioned!

**Comment by SanketDG** [comment]
@EfthymiaKostaki Can you resolve comments that you fixed? Otherwise it's hard to browse through every comment. You just have to click on the "Resolve comment" button if you have fixed it :)

**Comment by EfthymiaKostaki** [comment]
@SanketDG Sorry, I had not done this. I resolved everything except the new test which is not part of this issue so maybe you could resolve that and create a new issue? Since, you suggested.

**Review by vj-codes — CHANGES_REQUESTED** [review]
@EfthymiaKostaki hey please resolve the merge conflicts , until that's done this PR cannot be reviewed further.

**Comment by vj-codes** [comment]
@EfthymiaKostaki  closing this due to inactivity
Let us know if you would like to continue again and thank you for your contribution :)

**Pull request closed without merging**
